src/nbl/video/CVKLogicalDevice.h

```diff
@@ -124,51 +327,472 @@ class CVKLogicalDevice final : public ILogicalDevice
+        {
+            IDriverMemoryBacked::SDriverMemoryRequirements bufferMemoryReqs = {};
+
+            // These are only temporary
+            bufferMemoryReqs.vulkanBufferUsageFlags = initialMreqs.vulkanBufferUsageFlags;
+            bufferMemoryReqs.sharingMode = initialMreqs.sharingMode;
+            bufferMemoryReqs.queueFamilyIndices = initialMreqs.queueFamilyIndices;
+
+            // Not sure if I'd actually need these
+            bufferMemoryReqs.memoryHeapLocation = initialMreqs.memoryHeapLocation;
+            bufferMemoryReqs.mappingCapability = initialMreqs.mappingCapability;
+            bufferMemoryReqs.prefersDedicatedAllocation = initialMreqs.prefersDedicatedAllocation;
+            bufferMemoryReqs.requiresDedicatedAllocation = initialMreqs.requiresDedicatedAllocation;
+
+            vkGetBufferMemoryRequirements(m_vkdev, vk_buffer, &bufferMemoryReqs.vulkanReqs);
+
+            return core::make_smart_refctd_ptr<CVulkanBuffer>(
+                core::smart_refctd_ptr<CVKLogicalDevice>(this), bufferMemoryReqs, vk_buffer);
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+
+    core::smart_refctd_ptr<IGPUBuffer> createGPUBufferOnDedMem(const IDriverMemoryBacked::SDriverMemoryRequirements& initialMreqs, const bool canModifySubData = false) override
+    {
+        core::smart_refctd_ptr<IGPUBuffer> gpuBuffer = createGPUBuffer(initialMreqs, false);
+
+        if (!gpuBuffer)
+            return nullptr;
+
+        // Todo(achal): Probably do not call getMemoryReqs at all but
+        // vkGetBufferMemoryRequirements or equivalent
+        core::smart_refctd_ptr<video::IDriverMemoryAllocation> bufferMemory =
+            allocateDeviceLocalMemory(gpuBuffer->getMemoryReqs());
+
+        if (!bufferMemory)
+            return nullptr;
+
+        ILogicalDevice::SBindBufferMemoryInfo bindBufferInfo = {};
+        bindBufferInfo.buffer = gpuBuffer.get();
+        bindBufferInfo.memory = bufferMemory.get();
+        bindBufferInfo.offset = 0ull;
+        bindBufferMemory(1u, &bindBufferInfo);
+
+        return gpuBuffer;
+    }
+        
     core::smart_refctd_ptr<IGPUShader> createGPUShader(core::smart_refctd_ptr<asset::ICPUShader>&& cpushader) override
     {
-        return nullptr;
+        const asset::ICPUBuffer* source = cpushader->getSPVorGLSL();
+        core::smart_refctd_ptr<asset::ICPUBuffer> clone =
+            core::smart_refctd_ptr_static_cast<asset::ICPUBuffer>(source->clone(1u));
+        if (cpushader->containsGLSL())
+        {
+            return core::make_smart_refctd_ptr<CVulkanShader>(
+                core::smart_refctd_ptr<CVKLogicalDevice>(this), std::move(clone),
+                IGPUShader::buffer_contains_glsl);
+        }
+        else
+        {
+            return core::make_smart_refctd_ptr<CVulkanShader>(core::smart_refctd_ptr<CVKLogicalDevice>(this),
+                std::move(clone));
+        }
```

Comment:
ok in Vulkan, you'd hoist the compilation here, and drop any references to the `cpushader` source

`CVulkanShader` should only hold the driver's vkShaderModule

Reply:
Resolved

Reply:
compilation hoisted, but `CVulkanShader` still refcounts the SPIRV module source needlessly on your current branch @Erfan-Ahmadi